kubernetes/cache.go

```diff
@@ -41,18 +49,38 @@ type (
 		GetServices(namespace string) ([]core_v1.Service, error)
 		GetStatefulSet(namespace string, name string) (*apps_v1.StatefulSet, error)
 		GetStatefulSets(namespace string) ([]apps_v1.StatefulSet, error)
+		GetNamespaces() ([]core_v1.Namespace, error)
+
+		// Openshift caches
+		GetProjects() ([]osproject_v1.Project, error)
+
+		// Istio caches
+		GetVirtualServices(namespace string) (*GenericIstioObjectList, error)
+
+		// Allow subset of cache to be used
+		KubeCached(namespace string) bool
+		// OpenshiftCached(namespace string) bool
+		IstioCached(namespace string) bool
 	}
 
 	controllerImpl struct {
-		clientset       kube.Interface
-		refreshDuration time.Duration
-		stopChan        chan struct{}
-		syncCount       int
-		maxSyncCount    int
-		isErrorState    bool
-		lastError       error
-		lastErrorLock   sync.Mutex
-		controllers     map[string]cache.SharedIndexInformer
+		istioClient       IstioClient
+		clientset         *kube.Clientset
+		refreshDuration   time.Duration
+		stopChans         map[string]chan struct{}
+		controlStopChan   chan struct{}
+		syncCount         int
+		maxSyncCount      int
+		isErrorState      bool
+		lastError         error
+		lastErrorLock     sync.Mutex
+		controllers       map[string]map[string]cache.SharedIndexInformer
+		istioControllers  map[string]map[string]cache.SharedIndexInformer
+		projectInformer   cache.SharedIndexInformer
```

Comment:
This can be OpenShift projects or Kubernetes namespaces.
Perhaps a comment to miss the reference in the next methods.